cmd/piper.go

```diff
@@ -113,17 +112,65 @@ func addRootFlags(rootCmd *cobra.Command) {
 	rootCmd.PersistentFlags().BoolVar(&GeneralConfig.IgnoreCustomDefaults, "ignoreCustomDefaults", false, "Disables evaluation of the parameter 'customDefaults' in the pipeline configuration file")
 	rootCmd.PersistentFlags().StringVar(&GeneralConfig.ParametersJSON, "parametersJSON", os.Getenv("PIPER_parametersJSON"), "Parameters to be considered in JSON format")
 	rootCmd.PersistentFlags().StringVar(&GeneralConfig.EnvRootPath, "envRootPath", ".pipeline", "Root path to Piper pipeline shared environments")
-	rootCmd.PersistentFlags().StringVar(&GeneralConfig.StageName, "stageName", os.Getenv("STAGE_NAME"), "Name of the stage for which configuration should be included")
+	rootCmd.PersistentFlags().StringVar(&GeneralConfig.StageName, "stageName", "", "Name of the stage for which configuration should be included")
 	rootCmd.PersistentFlags().StringVar(&GeneralConfig.StepConfigJSON, "stepConfigJSON", os.Getenv("PIPER_stepConfigJSON"), "Step configuration in JSON format")
 	rootCmd.PersistentFlags().BoolVar(&GeneralConfig.NoTelemetry, "noTelemetry", false, "Disables telemetry reporting")
 	rootCmd.PersistentFlags().BoolVarP(&GeneralConfig.Verbose, "verbose", "v", false, "verbose output")
 	rootCmd.PersistentFlags().StringVar(&GeneralConfig.LogFormat, "logFormat", "default", "Log format to use. Options: default, timestamp, plain, full.")
 
 }
 
+const stageNameEnvKey = "STAGE_NAME"
+
+// initStageName initializes GeneralConfig.StageName from either GeneralConfig.ParametersJSON
+// or the environment variable 'STAGE_NAME', unless it has been provided as command line option.
+func initStageName(outputToLog bool) {
+	var stageNameSource string
+	if outputToLog {
+		defer func() {
+			log.Entry().Infof("Using stageName '%s' from %s", GeneralConfig.StageName, stageNameSource)
+		}()
+	}
+
+	if GeneralConfig.StageName != "" {
+		// Means it was given as command line argument and has the highest precedence
+		stageNameSource = "command line arguments"
+		return
+	}
+
+	// Use stageName from ENV as fall-back, for when extracting it from parametersJSON fails below
+	GeneralConfig.StageName = os.Getenv(stageNameEnvKey)
+	stageNameSource = fmt.Sprintf("env variable '%s'", stageNameEnvKey)
+
+	if len(GeneralConfig.ParametersJSON) == 0 {
+		return
+	}
+
+	var params map[string]interface{}
+	err := json.Unmarshal([]byte(GeneralConfig.ParametersJSON), &params)
+	if err != nil {
+		// Ignore here, will be logged later anyway
```

Comment:
Why? Where?

Reply:
During config resolution. [Here](https://github.com/SAP/jenkins-library/blob/master/pkg/config/config.go#L210).

Reply:
Not sure if we can always be sure that this will happen in the future. Would log it here again maybe with th additional info that the the stage name might be ignored.